### ISSUE TITLE
Why Model YOLOv8x-obb.pt is only be 59.8% mAPtest50 tested on DOTAv1 task1 leaderboard, which is quite different from 81.36% in README? How can I reach 81.36% in DOTAv1 task1 leaderboard?

### ISSUE THREAD

**Issue by xxxyyynnn** [ultralytics/ultralytics#9262]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Why Model YOLOv8x-obb.pt is only be 59.8% mAPtest50 tested on DOTAv1 task1 leaderboard, which is quite different from 81.36% in README? How can I reach 81.36% in DOTAv1 task1 leaderboard?
![aec6440c16751a46d81418015154c5a](https://github.com/ultralytics/ultralytics/assets/92631419/25a6566f-0ca6-4a92-a379-d4471267aa94)

**Comment by xxxyyynnn** [comment]
I feel quite confused. Could you please help me to reach 81.36% in DOTAv1 task1 leaderboard?

**Comment by glenn-jocher** [comment]
@xxxyyynnn absolutely, we're here to help! 🌟 The difference in performance (59.8% vs. 81.36% mAPtest50) might be attributed to several factors including the model configuration, dataset pre-processing, or evaluation method. To aim for the higher mAP, it's crucial to ensure the training setup closely follows the one described for achieving the 81.36% result. This includes:

- Using the exact version of the model (`YOLOv8x-obb`).
- Ensuring the dataset is prepared and pre-processed as mentioned for the DOTAv1 task1.
- Applying the same hyperparameters and training configurations.

It's also beneficial to review the training logs for any indications of divergences in expected behavior or potential issues. 

If you've followed these steps and the issue persists, could you share more details about your training setup and configuration? This way, we can provide more tailored advice. 🛠️

**Comment by xxxyyynnn** [comment]
I've followed these steps, but the issue persists. And the details about my predict setup are as follows:

1. I download yolov8x-obb.pt file from the link "https://github.com/ultralytics/assets/releases/download/v8.1.0/yolov8x-obb.pt", and put it in path "/home/xyn02/ultralytics/yolov8x-obb.pt". I download the DOTAv1 dataset from the link "https://github.com/ultralytics/yolov5/releases/download/v1.0/DOTAv1.zip", and unzip it and put the test images in path "/home/xyn02/ultralytics/ultralytics/datasets/DOTAv1/images/test".

2. I run the following code.

from ultralytics import YOLO
# Load a pretrained model
model = YOLO('/home/xyn02/ultralytics/yolov8x-obb.pt')
model.predict('/home/xyn02/ultralytics/ultralytics/datasets/DOTAv1/images/test', save=True, save_txt=True, save_conf=True, show_labels=False, show_conf=False)

3. I could get the predicted labels of DOTAv1 test sets in "/home/xyn02/ultralytics/runs/obb/predict/labels".

4. I pre-processed the labels in "/home/xyn02/ultralytics/runs/obb/predict/labels", and I'm sure I have correctly converted to DOTAv1 task1 format. The convert code is as follows.

import os
from collections import defaultdict
from PIL import Image


def denormalize_coordinates(normed_coordinates, image_width, image_height):
    denormed_coordinates = []
    for coord in normed_coordinates:
        x = coord[0] * image_width
        y = coord[1] * image_width
        denormed_coordinates.append((x, y))
    return denormed_coordinates


def process_txt_files(image_folder, label_folder, output_folder):
    class_mapping = {
        0: "plane",
        1: "ship",
        2: "storage tank",
        3: "baseball diamond",
        4: "tennis court",
        5: "basketball court",
        6: "ground track field",
        7: "harbor",
        8: "bridge",
        9: "large vehicle",
        10: "small vehicle",
        11: "helicopter",
        12: "roundabout",
        13: "soccer ball field",
        14: "swimming pool"
    }

    class_data = defaultdict(list)

    for txt_file in os.listdir(label_folder):
        if txt_file.endswith(".txt"):
            image_name = os.path.splitext(txt_file)[0]  # Remove the file extension
            image_path = os.path.join(image_folder, image_name + ".jpg")  # Add the file extension back
            label_path = os.path.join(label_folder, txt_file)

            with open(label_path, 'r') as f:
                lines = f.readlines()

            for line in lines:
                line = line.strip().split()
                class_num = int(line[0])
                confidence = float(line[-1])  # Confidence is the last value in the line
                normed_coordinates = [(float(line[i]), float(line[i+1])) for i in range(1, len(line)-1, 2)]  # Exclude confidence
                # Get image dimensions
                image = Image.open(image_path)
                image_width, image_height = image.size
                # Denormalize coordinates
                denormed_coordinates = denormalize_coordinates(normed_coordinates, image_width, image_height)
                # Append to class_data
                class_data[class_num].append((image_name, confidence, denormed_coordinates))

    # Write class data to files
    for class_num, data in class_data.items():
        class_name = class_mapping[class_num]
        class_name = class_name.replace(" ", "-")  # Replace space with hyphen
        output_path = os.path.join(output_folder, f"Task1_{class_name}.txt")
        with open(output_path, 'w') as f:
            for image_name, confidence, coordinates in data:
                f.write(f"{image_name} {confidence} {' '.join(map(str, sum(coordinates, ())))}\n")


if __name__ == "__main__":
    image_dir = "/home/xyn02/ultralytics/ultralytics/datasets/DOTAv1/images/test"
    label_dir = "/home/xyn02/ultralytics/runs/obb/predict/labels"
    output_dir = "/home/xyn02/ultralytics/Leaderboard/predict_labelLeaderboard"

    os.makedirs(output_dir, exist_ok=True)

    process_txt_files(image_dir, label_dir, output_dir)

5. I zipped the file "/home/xyn02/ultralytics/Leaderboard/predict_labelLeaderboard" and upload the "predict_labelLeaderboard.zip" to the link "http://[bed4rs.net:8001/evaluation1/](http://bed4rs.net:8001/evaluation1/)".

6. But I only get the following result.

This is your evaluation result for task 1 (VOC metrics):

mAP: 0.3514136347707364
ap of each class: plane:0.5376300505996345, baseball-diamond:0.4408032021358564, bridge:0.0606060606060606, ground-track-field:0.3276236044657097, small-vehicle:0.33678242431606087, large-vehicle:0.6618614665273004, ship:0.3418597502228976, tennis-court:0.8142485739523534, basketball-court:0.44270833333333337, storage-tank:0.35415967487470523, soccer-ball-field:0.3085450264021693, roundabout:0.15087040618955513, harbor:0.311687766117228, swimming-pool:0.09090909090909091, helicopter:0.09090909090909091
COCO style result:

AP50: 0.3514136347707364
AP75: 0.1949671106131959
mAP: 0.20963523821874536

7. I feel quite confused. Could you please help me tackle the problem that have troubled me several days?

**Comment by glenn-jocher** [comment]
@xxxyyynnn hey there! 🌞 Thanks for sharing the detailed steps of your process. It’s great to see you've carefully set up everything for the DOTAv1 task1 validation. Based on the description and the results, there are a couple of points we might consider to improve the performance:

- **Model and Dataset Compatibility:** Ensure that the YOLOv8x-obb model you're using is specifically optimized for the DOTAv1 dataset. Sometimes, models might require fine-tuning on a specific dataset to achieve the best results.

- **Preprocessing and Postprocessing Matching:** Confirm that the image preprocessing during prediction matches the preprocessing used during the model training phase. This includes image normalization, resizing methods, etc. Similarly, make sure your postprocessing (conversion to DOTAv1 format) aligns with any standards or expectations of the evaluation server.

- **Label Formatting:** Double-check the converted label format against the DOTAv1 task1 specification. Even slight deviations in the format or misinterpretations of coordinates could significantly affect the evaluation metrics.

Given everything looks correct on your end, it might be helpful to directly reach out on the evaluation platform or forums related to the DOTAv1 dataset for any known issues or particularities with the evaluation process that could be affecting your scores.

If possible, reviewing or reaching out to others who successfully achieved higher scores could offer some insights. They might have encountered and solved similar issues during their attempts. 🚀

Let’s keep the conversation going, and feel free to share any updates or findings!

**Comment by xxxyyynnn** [comment]
1. For model and dataset capacity, I can't figure out what the fine-tunning is when you yolov8-obb model. Could you please tell me?


2. And I don't know how you preprocess and postprocess the dataset. I just use the original test set of DOTAv1 to predict without any proprocess. Could you please teach me?


3. About label Formatting, I have double checked the converted label format, and my labels are only different in the number of decimal places reserved with the example from DOTAv1 official website. Some of my labels are as follows.


P0045 0.925 602.4451250000001 1500.66075 852.36435 1508.581025 844.5312250000001 1755.7239 594.6140750000001 1747.803625



---Original---
From: "Glenn ***@***.***&gt;
Date: Mon, Mar 25, 2024 06:34 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] Why Model YOLOv8x-obb.pt is only be59.8% mAPtest50 tested on DOTAv1 task1 leaderboard, which is quite differentfrom 81.36% in README? How can I reach 81.36% in DOTAv1 task1 leaderboard?(Issue #9262)




 
@xxxyyynnn hey there! 🌞 Thanks for sharing the detailed steps of your process. It’s great to see you've carefully set up everything for the DOTAv1 task1 validation. Based on the description and the results, there are a couple of points we might consider to improve the performance:
  
 
Model and Dataset Compatibility: Ensure that the YOLOv8x-obb model you're using is specifically optimized for the DOTAv1 dataset. Sometimes, models might require fine-tuning on a specific dataset to achieve the best results.
 
 
 
Preprocessing and Postprocessing Matching: Confirm that the image preprocessing during prediction matches the preprocessing used during the model training phase. This includes image normalization, resizing methods, etc. Similarly, make sure your postprocessing (conversion to DOTAv1 format) aligns with any standards or expectations of the evaluation server.
 
 
 
Label Formatting: Double-check the converted label format against the DOTAv1 task1 specification. Even slight deviations in the format or misinterpretations of coordinates could significantly affect the evaluation metrics.
 
  
Given everything looks correct on your end, it might be helpful to directly reach out on the evaluation platform or forums related to the DOTAv1 dataset for any known issues or particularities with the evaluation process that could be affecting your scores.
 
If possible, reviewing or reaching out to others who successfully achieved higher scores could offer some insights. They might have encountered and solved similar issues during their attempts. 🚀
 
Let’s keep the conversation going, and feel free to share any updates or findings!

**Comment by xxxyyynnn** [comment]
It's pity that though I have round the coordinates as the example provided, the mAP is still low, just 0.5859024894980936.

P0045 0.925 602 1501 852 1509 845 1756 595 1748
P0045 0.916802 680 1890 929 1909 909 2153 660 2133
P0045 0.914932 998 1428 1247 1432 1243 1681 994 1677
P0045 0.894974 483 2361 734 2365 731 2615 479 2611
P0045 0.878815 1074 1826 1322 1834 1314 2080 1067 2073
P0046 0.939422 248 736 419 739 415 906 245 903
P0046 0.912001 1148 1676 1277 1678 1275 1807 1146 1806
P0046 0.891353 910 1439 1034 1441 1033 1565 909 1564
P0046 0.872375 820 604 981 606 978 773 817 771
P0048 0.901609 878 841 1058 846 1053 1021 873 1016
P0048 0.874191 1099 1186 1309 1210 1286 1406 1076 1382

**Comment by xxxyyynnn** [comment]
I use the exact version of the model (YOLOv8x-obb), but the result is bad.

![P0048](https://github.com/ultralytics/ultralytics/assets/92631419/f5c0bdb3-6bd9-4c36-be66-506a63ab0488)

**Comment by glenn-jocher** [comment]
@xxxyyynnn hey there! 😊 It looks like you're facing some issues despite using the exact model version. Sometimes, discrepancies can arise from various subtle factors. To help you troubleshoot, could you ensure:

- **Data Preprocessing:** The test data should be preprocessed (normalized, resized, etc.) in the exact manner as during training.
- **Inferencing Parameters:** Double-check that the inference parameters (like confidence threshold, NMS threshold) match those used during your model's evaluation phase or recommendation.

Here's a quick check for inferencing parameters:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('/path/to/yolov8x-obb.pt')

# Make predictions with recommended inference settings
model.predict('/path/to/images', conf=0.25, iou=0.45)
```

If these settings still don't improve the results, it might be worth revisiting the training logs or experimenting with slight adjustments to these parameters. Let's get this sorted! 🛠️

**Comment by xxxyyynnn** [comment]
@glenn-jocher Could you please offer me all of the parameters that you use? Maybe the parameters that ruined my results.

**Comment by xxxyyynnn** [comment]
@glenn-jocher Hello! I want to figure out how you can achieve mAPtest50=81.36. Could you please offer me all of the parameters that you use? Maybe the parameters that ruined my results.

**Comment by xxxyyynnn** [comment]
@glenn-jocher Hello! I can't understand why mAPtest50 is reproduced by val(yolo val obb data=DOTAv1.yaml device=0 split=test). Could you please help me?

![image](https://github.com/ultralytics/ultralytics/assets/92631419/618d598c-4092-4011-a151-83a7c988a9cd)

**Comment by chillaxshine** [comment]
I have a similar problem. I checked the picture of P0048 and found that the rotation box of baseball was not correct. My code is
            model = YOLO('yolov8x-obb.pt')
            model('/home/sxx/DOTAv1/images/test'， save=True)
 I just git clone  the code without any changes.
![P0048](https://github.com/ultralytics/ultralytics/assets/156276177/919f0d72-3005-4a23-8d2b-d6967f285c9d)

**Comment by glenn-jocher** [comment]
@chillaxshine hey there! 😊 Seeing issues with object detection, especially with rotated bounding boxes like in your P0048 example, can indeed be puzzling. It looks like your setup is straightforward, which is great. Sometimes, discrepancies in detection, especially with orientation, can stem from how the model was trained or slight nuances in the dataset preparation.

A good first step is to double-check your dataset's annotations and preprocessing steps to ensure they're aligned with how the model expects them. Additionally, considering updating your model or re-downloading it to ensure there's no corruption might help:

```python
from ultralytics import YOLO

# Load and run predictions with the model
model = YOLO('path/to/yolov8x-obb.pt')
results = model.predict('/home/sxx/DOTAv1/images/test', save=True)
```

Make sure the path to your model and the dataset is correct (`path/to/yolov8x-obb.pt` should be replaced with the actual file path). If the issue persists, doing a deep dive into the specific cases where the rotation seems off and comparing them against your dataset might shed some light. 

Keep us posted on how it goes or if you discover anything new! 👍

**Comment by chillaxshine** [comment]
@glenn-jocher I git clone again and get a new code and executed the code you mentioned, but the result is still the same. Could you please provide the predicted result of Picture P0048? I need this picture and want to confirm whether there is a problem, thank you!

**Comment by glenn-jocher** [comment]
@chillaxshine hey there! 😊 I appreciate you giving it another go with a fresh clone. It sounds like you're keen to deep-dive into this specific case with Picture P0048. Unfortunately, I can't directly provide predicted results, but here's a helping hand to guide you towards generating them yourself:

```python
from ultralytics import YOLO

# Make sure to adjust the path to where your model is located
model = YOLO('path/to/yolov8x-obb.pt') 

# Now, replace 'path/to/P0048.jpg' with the actual path to your image
results = model.predict('path/to/P0048.jpg')

# This will save the prediction in the runs/detect folder
results.show()  # Optionally display the image with predictions
```

This should help you generate the predictions for P0048 and verify them. If there's still a discrepancy or issue, it might be worth comparing the annotation formats or even trying a different image to pinpoint the problem. 

Let's keep the conversation going. Always here to help!

**Comment by Doan-IT** [comment]
@chillaxshine It seems like he used LLM to answer the question so the content was quite general and meaningless.
